app/memorize.py

```diff
@@ -0,0 +1,127 @@
+from functools import wraps
+import hashlib
+import pickle
+
+from wrapt import ObjectProxy
+
+
+class CacheMissError(RuntimeError):
+    """
+    Exception thrown when memorized object is not found in cache and
+    error_on_miss is set to True.
+    """
+    pass
+
+
+class MemorizedObject(ObjectProxy):
+    """
+    Thin wrapper around any memorized objects, adding attributes indicating
+    the object's cache key and whether or not it was returned from cache.
+    """
+    def __init__(self, wrapped):
+        super(MemorizedObject, self).__init__(wrapped)
+        self._self_from_cache = False
+        self._self_cache_key = None
+
+    def __reduce__(self):
+        return type(self), (self.__wrapped__,)
```

Comment:
Should add docstring indicating that this is to make `MemorizedObject` instances pickable.